optax/contrib/_sophia_h.py

```diff
@@ -0,0 +1,307 @@
+      free, theoretically increasing hessian estimation accuracy.
+
+  References:
+    Liu et al., `Sophia: A Scalable Stochastic Second-order Optimizer for
+    Language Model Pre-training <https://arxiv.org/abs/2305.14342>`_, 2023
+
+    `Levanter <https://www.github.com/stanford-crfm/levanter>`_
+
+  Args:
+    b1: Exponential decay rate for the first moment estimates.
+    b2: Exponential decay rate for the hessian diagonal estimates. Keep in mind
+        effective `b2` is `1 - (1 - b2) / update_interval`, e.g. default `b2`
+        of 0.99 is effectively 0.999 because default `update_interval` is every
+        10.
+    eps: Small constant to avoid division by zero.
+    gamma: Normalizing constant for the hessian diagonal.
+    clip_threshold: Threshold for clipping updates.
+    update_interval: Interval for updating the hessian diagonal.
+    mu_dtype: dtype of the first moment estimates.
+    pmap_axis_name: Provide pmap axis name if using pmap to perform separate
+        monte carlo samples on each device for hutchinson's estimator for
+        (almost) free.
+    seed: PRNGKey.
+    print_win_rate_every_n_steps: Print sophia win rate every n steps for
+        diagnostic purposes. Authors state this value should stay between
+        0.1 and 0.5 during training. If win rate is too low, try increasing
+        `gamma`.
+
+  Returns:
+    optax.GradientTransformationExtraArgs
+  """
+  mu_dtype = canonicalize_dtype(mu_dtype)
+
+  def init_fn(params):
+    mu = jax.tree.map(lambda t: jnp.zeros_like(t, dtype=mu_dtype), params)
+    nu = jax.tree.map(jnp.zeros_like, params)
+    key = seed
+    if pmap_axis_name and jax.local_device_count() > 1:
+      key = jax.random.split(key, jax.local_device_count())
+    return SophiaHState(
+        count=jnp.zeros([], jnp.int32), mu=mu, nu=nu, key=key
+    )
+
+  def update_fn(
+      updates, state: SophiaHState, params=None, obj_fn=None
+  ):
+    if params is None:
+      raise ValueError("params must be provided to sophia's update function.")
+    if obj_fn is None:
+      raise ValueError(
+          "obj_fn must be provided to sophia's update function. "
+          "See optimizer docstring for more information."
+      )
+    count_inc = safe_int32_increment(state.count)
+
+    mu = otu.tree_update_moment(updates, state.mu, b1, 1)
+    mu_hat = otu.tree_bias_correction(mu, b1, count_inc)
+    updates = jax.tree.map(
+        lambda m, h: m / jnp.maximum(gamma * h, eps), mu_hat, state.nu
+    )
+    if clip_threshold is not None:
+      sum_not_clipped = jax.tree.reduce(
+          lambda x, y: x + y,
+          jax.tree.map(
+              lambda u: jnp.sum(jnp.abs(u) < clip_threshold), updates
+          ),
+      )
+      total_tree_size = sum(x.size for x in jax.tree.leaves(updates))
+      win_rate = sum_not_clipped / total_tree_size
+      jax.lax.cond(
+          jnp.logical_and(
+              print_win_rate_every_n_steps > 0,
```

Comment:
I would wrap the logic associated to ``print_win_rate_every_n_steps > 0`` in an if statement to keep clear what operations are necessary for the algorithm and what are operations are simply for debugging if ``print_win_rate_every_n_steps > 0``

Reply:
I did have it that way but changed it to this to pass the injected hyperparams test, but if it's ok it fails that test I'll switch it back to the if statement. I could also modify the inject hyperparams test (in contrib._common_test.py) to not check every argument, just let me know what would be best

Reply:
Ah I see. Let's keep it the way you had it then.